### PR TITLE
[3.3.5] Azjol-Nerub rewrite

### PR DESCRIPTION
**Pull request by Treeston**
Starting with Azjol-Nerub because you have to start somewhere (and Hadronox not working annoys me).

This is obviously heavily WIP. Comments and feedback are appreciated, though only on parts that have been newly added and/or modified by the current version of this PR. Please do not comment on creatures/encounters that have not yet been touched.

Progress:
- [x] Krik'thir the Gatewatcher
- [x] Hadronox
- [x] Anub'arak
- [x] Assorted cleanup

### CODE REVIEW

File: sql/updates/world/3.3.5/9999_99_99_99_AZJOLNERUB.sql [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+-- Krik'thir the Gatewatcher --
+-- ------------------------- --
+DELETE FROM `conditions` WHERE `SourceTypeOrReferenceId`=13 AND `SourceEntry` IN (52438,52449);
+INSERT INTO `conditions` (`SourceTypeOrReferenceId`,`SourceGroup`,`SourceEntry`,`SourceId`,`ElseGroup`,`ConditionTypeOrReference`,`ConditionTarget`,`ConditionValue1`,`ConditionValue2`,`ConditionValue3`,`NegativeCondition`,`Comment`) VALUES
+(13,1,52438,0,0,31,0,3,22515,0,0,"Summon Skittering Swarmer - Target World Trigger"),
+(13,1,52449,0,0,31,0,3,22515,0,0,"Summon Skittering Infector - Target World Trigger");
+
+
+-- Anub'ar Skirmisher: Fixtate
+DELETE FROM `spell_script_names` WHERE `spell_id`=52536;
```
Comment (Keader) [inline sql/updates/world/3.3.5/9999_99_99_99_AZJOLNERUB.sql]: tc always delete script by name

**Comment by sirikfoll** [comment]
Personal opinion, please feel free to ignore or delete.
What do you think about using TaskScheduler instead of EventMap? I think it looks "cleaner" in smaller scripts, like those from Dungeons

**Comment by Treeston** [comment]
Never worked with it much, admittedly.

File: src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_krikthir_the_gatewatcher.cpp [hunk 2]
```diff
@@ -88,386 +117,374 @@ class boss_krik_thir : public CreatureScript
...
 
-            void JustDied(Unit* /*killer*/) override
+            while (uint32 eventId = _events.ExecuteEvent())
             {
-                Talk(SAY_DEATH);
-                _JustDied();
+                switch (eventId)
+                {
+                    case EVENT_ENRAGE:
+                        DoCast(nullptr, SPELL_ENRAGE);
```
Comment (Keader) [inline src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_krikthir_the_gatewatcher.cpp]: this "nullptr" is necessary?
can be  DoCast(SPELL_ENRAGE); , no?

Comment (Treeston) [inline src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_krikthir_the_gatewatcher.cpp]: `DoCast` with a single argument is a [pretty weird overload](https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/AI/CoreAI/UnitAI.cpp#L135).

**Comment by ccrs** [comment]
joschiwald had some workaround :)

File: src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_krikthir_the_gatewatcher.cpp [hunk 3]
```diff
@@ -116,7 +116,31 @@ class boss_krik_thir : public CreatureScript
...
         {
-            boss_krik_thirAI(Creature* creature) : BossAI(creature, DATA_KRIKTHIR_THE_GATEWATCHER) { }
+            public:
+                boss_krik_thirAI(Creature* creature) : BossAI(creature, DATA_KRIKTHIR_THE_GATEWATCHER), _hadGreet(false) { }
+
+                void DoAction(int32 action) override
+                {
+                    switch (action)
+                    {
+                        case -ACTION_GATEWATCHER_GREET:
```
Comment (P-Kito) [inline src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_krikthir_the_gatewatcher.cpp]: Negative intended? What does it do? Never seen it before.

Comment (Treeston) [inline src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_krikthir_the_gatewatcher.cpp]: I use negative values for actions defined in the instance script's header, and positive values for actions originating in the boss script.

File: sql/updates/world/3.3.5/9999_99_99_99_AZJOLNERUB.sql [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+(13,1,52449,0,0,31,0,3,22515,0,0,"Summon Skittering Infector - Target World Trigger");
+
+-- Anub'ar Skirmisher: Fixtate
+DELETE FROM `spell_script_names` WHERE `scriptname` = "spell_anub_ar_skirmisher_fixtate";
+INSERT INTO `spell_script_names` (`spell_id`,`scriptname`) VALUES (52536,"spell_anub_ar_skirmisher_fixtate");
+
+-- Watchers: Web Wrap 
+DELETE FROM `spell_script_names` WHERE `scriptname` = "spell_gatewatcher_web_wrap";
+INSERT INTO `spell_script_names` (`spell_id`,`scriptname`) VALUES (52086,"spell_gatewatcher_web_wrap");
+UPDATE `creature_template` SET `AIName`='',`ScriptName`="npc_gatewatcher_web_wrap" WHERE `entry`=28619;
```
Comment (Keader) [inline sql/updates/world/3.3.5/9999_99_99_99_AZJOLNERUB.sql]: missing new line

**Comment by joschiwald** [comment]
about hadronox: https://github.com/joschiwald/TrinityCore/tree/hadronox

File: src/server/scripts/Northrend/AzjolNerub/AzjolNerub/azjol_nerub.h [hunk 5]
```diff
@@ -55,10 +59,4 @@ enum GameObjectIds
     GO_ANUBARAK_DOOR_3              = 192398
 };
 
-template<class AI>
-AI* GetAzjolNerubAI(Creature* creature)
-{
-    return GetInstanceAI<AI>(creature, AzjolNerubScriptName);
-}
```
Comment (joschiwald) [inline src/server/scripts/Northrend/AzjolNerub/AzjolNerub/azjol_nerub.h]: why did you delete it?

File: src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_krikthir_the_gatewatcher.cpp [hunk 6]
```diff
@@ -61,122 +110,275 @@ enum Yells
...
+            return GetInstanceAI<boss_krik_thirAI>(creature);
+        }
 };
 
-class boss_krik_thir : public CreatureScript
+class npc_watcher_gashra : public CreatureScript
 {
     public:
-        boss_krik_thir() : CreatureScript("boss_krik_thir") { }
+    npc_watcher_gashra() : CreatureScript("npc_watcher_gashra") { }
```
Comment (joschiwald) [inline src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_krikthir_the_gatewatcher.cpp]: bad indent

File: src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_krikthir_the_gatewatcher.cpp [hunk 7]
```diff
@@ -61,122 +110,275 @@ enum Yells
...
-                            events.ScheduleEvent(EVENT_CURSE_FATIGUE, 10000);
-                            break;
-                        default:
-                            break;
-                    }
-                }
+class npc_watcher_narjil : public CreatureScript
+{
+    public:
+    npc_watcher_narjil() : CreatureScript("npc_watcher_narjil") { }
```
Comment (joschiwald) [inline src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_krikthir_the_gatewatcher.cpp]: bad indent

File: src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_krikthir_the_gatewatcher.cpp [hunk 8]
```diff
@@ -61,220 +118,303 @@ enum Yells
...
-
 class boss_krik_thir : public CreatureScript
 {
     public:
         boss_krik_thir() : CreatureScript("boss_krik_thir") { }
 
         struct boss_krik_thirAI : public BossAI
         {
-            boss_krik_thirAI(Creature* creature) : BossAI(creature, DATA_KRIKTHIR_THE_GATEWATCHER) { }
+            public:
```
Comment (joschiwald) [inline src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_krikthir_the_gatewatcher.cpp]: public by default, remove this keyword to keep consistence with all other scripts

**Comment by Treeston** [comment]
Hadronox almost done. Review appreciated.

**Comment by Treeston** [comment]
Hadronox done. Please test and review.

File: sql/updates/world/3.3.5/9999_99_99_99_AZJOLNERUB.sql (hunk 4 above)
Comment (tkrokli) [inline sql/updates/world/3.3.5/9999_99_99_99_AZJOLNERUB.sql]: Fixed after comment above (but comment was after the EOF and was not hidden by the update).

File: src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_krikthir_the_gatewatcher.cpp [hunk 9]
```diff
@@ -795,6 +814,8 @@ class npc_skittering_swarmer : public CreatureScript
             {
                 ScriptedAI::InitializeAI();
                 DoZoneInCombat();
+                if (Creature* gatewatcher = me->GetInstanceScript()->GetCreature(DATA_KRIKTHIR_THE_GATEWATCHER))
+                    gatewatcher->AI()->SetGUID(me->GetGUID());
```
Comment (joschiwald) [inline src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_krikthir_the_gatewatcher.cpp]: what about using JustSummoned hook?

File: src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_krikthir_the_gatewatcher.cpp [hunk 10]
```diff
@@ -630,15 +793,204 @@ class npc_watcher_silthik : public CreatureScript
...
+        struct npc_skittering_swarmerAI : public ScriptedAI
+        {
+            npc_skittering_swarmerAI(Creature* creature) : ScriptedAI(creature) { }
+
+            void InitializeAI() override
+            {
+                ScriptedAI::InitializeAI();
+                DoZoneInCombat();
+                if (Creature* gatewatcher = me->GetInstanceScript()->GetCreature(DATA_KRIKTHIR_THE_GATEWATCHER))
+                    gatewatcher->AI()->SetGUID(me->GetGUID());
```
Comment (joschiwald) [inline src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_krikthir_the_gatewatcher.cpp]: what about using JustSummoned hook?

Comment (Treeston) [inline src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_krikthir_the_gatewatcher.cpp]: It's owned by the world trigger that gets hit by the spell, not Gatewatcher himself.

Comment (sirikfoll) [inline src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_krikthir_the_gatewatcher.cpp]: I think he meant something like `gatewatcher->AI()->JustSummoned(me);` , so you can use `summons.DespawnEntry()` to despawn them later

Comment (Treeston) [inline src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_krikthir_the_gatewatcher.cpp]: For the Hadronox case, I don't want them in the summon list specifically, as summon list has some specific behavior like assist on aggro etc.

For Gatewatcher case, that's a good point.

**Comment by Treeston** [comment]
Initial version of Anub'arak encounter done. Testing and feedback appreciated.

**Comment by Treeston** [comment]
Current issues:
[ ] Anub'rekhan spike visuals spawned on all targets (needs limitation to 1 target)
[ ] Anub'rekhan wall goes up too quickly (delayed on retail)

**Comment by Nawuko** [comment]
And maybe check via sniffs if the submerge is timed or after all add wave gets killed.

**Comment by tkrokli** [comment]
@Nawuko : surface / ascend / rise / float ? (antonyms of 'submerge')

**Comment by ccrs** [comment]
The wall goes up on _EnterCombat();

The original script avoided the call (adding the inside calls manually) and used a timer.

**Comment by Ascathor** [comment]
Does he gain immunity to effects on precast of submerge ?

File: src/server/scripts/Northrend/AzjolNerub/AzjolNerub/azjol_nerub.h [hunk 11]
```diff
@@ -18,22 +18,22 @@
...
-#define AzjolNerubScriptName "instance_azjol_nerub"
 #define DataHeader           "AN"
 
 uint32 const EncounterCount = 3;
 
 enum DataTypes
 {
     // Encounter States/Boss GUIDs
-    DATA_KRIKTHIR_THE_GATEWATCHER   = 0,
+    DATA_KRIKTHIR   = 0,
```
Comment (Keader) [inline src/server/scripts/Northrend/AzjolNerub/AzjolNerub/azjol_nerub.h]: codestyle

File: src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_anubarak.cpp [hunk 12]
```diff
@@ -52,320 +60,639 @@ enum Yells
...
-        uint32 DatterTimer;
-        uint32 DelayTimer;
-
-        uint32 ImpaleTimer;
-        uint32 ImpalePhase;
-        ObjectGuid ImpaleTarget;
+        boss_anub_arakAI(Creature* creature) : BossAI(creature, DATA_ANUBARAK), _nextSubmerge(0), _petCount(0), _assassinCount(0), _guardianCount(0), _venomancerCount(0) { }
 
-        SummonList Summons;
+        void Initialize() { }
```
Comment (Keader) [inline src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_anubarak.cpp]: if Initialize is empty, why you keep?

File: src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_hadronox.cpp [hunk 13]
```diff
@@ -15,192 +15,1084 @@
...
+{
+    DATA_CRUSHER_PACK_ID = 1,
+    DATA_HADRONOX_WEBBED_DOORS
+};
+
+enum Creatures
+{
+    NPC_CRUSHER = 28922,
+    NPC_WORLDTRIGGER_LARGE = 23472
+};
```
Comment (Keader) [inline src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_hadronox.cpp]: identation

**Comment by Treeston** [comment]
Initial CS/cleanup pass done. Review and testing appreciated.

File: src/server/game/Instances/InstanceScript.cpp [hunk 14]
```diff
@@ -288,34 +288,34 @@ bool InstanceScript::SetBossState(uint32 id, EncounterState state)
 {
     if (id < bosses.size())
     {
```
Comment (Shauren) [inline src/server/game/Instances/InstanceScript.cpp]: all changes in this file are unneeded refactoring

File: src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_anubarak.cpp [hunk 15]
```diff
@@ -21,351 +21,673 @@
...
-                        if (Creature* impaleTarget = ObjectAccessor::GetCreature(*me, ImpaleTarget))
-                            me->CastSpell(impaleTarget, SPELL_IMPALE_DMG, true);
-                        ImpalePhase = IMPALE_PHASE_TARGET;
-                        ImpaleTimer = 9*IN_MILLISECONDS;
+                    case EVENT_GUARDIAN:
+                        if (Creature* trigger = ObjectAccessor::GetCreature(*me, _guardianTrigger))
+                        {
+                            trigger->CastSpell(trigger, SPELL_SUMMON_GUARDIAN, true);
+                            trigger->CastSpell(trigger, SPELL_SUMMON_GUARDIAN, true);
+                            if (_guardianCount -= 2)
```
Comment (Shauren) [inline src/server/scripts/Northrend/AzjolNerub/AzjolNerub/boss_anubarak.cpp]: please stop doing assignments inside conditional statements if they do not declare the variable in the same place as well `if (Type var = something)` is ok but `if (pre_existing_var = something)` is bad

**Comment by Treeston** [comment]
TODO rewrite hadronox add movements using smooth pathing.
